Remora.Rest/Json/Internal/BoundDataObjectConverter.cs

```diff
@@ -0,0 +1,272 @@
+            .ToDictionary(x => x.n, x => (x.IsPrimary, x.DTOProperty));
+    }
+
+    private static JsonSerializerOptions GetPropertyConverterOptions(JsonSerializerOptions options, JsonConverter? converter)
+    {
+        if (converter == null)
+        {
+            return options;
+        }
+
+        var cloned = new JsonSerializerOptions(options);
+        cloned.Converters.Insert(0, converter);
+
+        return cloned;
+    }
+
+    /// <inheritdoc />
+    public override TInterface Read
+    (
+        ref Utf8JsonReader reader,
+        Type typeToConvert,
+        JsonSerializerOptions options
+    )
+    {
+        if (reader.TokenType != JsonTokenType.StartObject ||
+            !reader.Read())
+        {
+            DataObjectConverterShared.ThrowJsonException();
+        }
+
+        var readProperties = _readProperties;
+
+        // CMBK: Considering to cache this array and reuse it.
+        // F.e. in a thread-static field. This method isn't async anyways.
+
+        // Avoid creating a dictionary and fill the arguments array directly
+        var constructorArguments = new object?[readProperties.Length];
+
+        // We use "Missing" as a sentinel to indicate that a value hasn't
+        // been seen yet without needing to allocate additional memory.
+        constructorArguments.AsSpan().Fill(DataObjectConverterShared.Missing);
+
+        while (reader.TokenType != JsonTokenType.EndObject)
+        {
+            var propertyName = reader.GetString();
+            if (!reader.Read())
+            {
+                DataObjectConverterShared.ThrowJsonException();
+            }
+
+            Debug.Assert(propertyName != null, "Expected property names to never be null.");
+            var (isPrimaryChoice, dtoProperty) = GetReadPropertyInfo(propertyName);
+
+            if (dtoProperty is null)
+            {
+                if (!_allowExtraProperties)
+                {
+                    DataObjectConverterShared.ThrowJsonException();
+                }
+
+                // No matching property - we'll skip it
+                if (!reader.TrySkip())
+                {
+                    DataObjectConverterShared.ThrowJsonException("Couldn't skip elements.");
+                }
+
+                if (!reader.Read())
+                {
+                    DataObjectConverterShared.ThrowJsonException
+                    (
+                        $"No matching DTO property for JSON property \"{propertyName}\" could be found."
+                    );
+                }
+
+                continue;
+            }
+
+            var propertyValue = JsonSerializer.Deserialize(ref reader, dtoProperty.Property.PropertyType, dtoProperty.Options);
+
+            // Verify nullability
+            if (propertyValue is null && !dtoProperty.AllowsNull)
+            {
+                DataObjectConverterShared.ThrowJsonException();
```

Comment:
Add explanatory error message.

Reply:
https://github.com/Remora/Remora.Rest/blob/cfc35dafd76ba3732406714c94685f4fb712e9c1/Remora.Rest/Json/Internal/BoundDataObjectConverter.cs#L199
Should be fine?

Reply:
I prefer throw helper with the ability to pass in that message into it instead (throw helpers help with inline by the JIT).

Reply:
That's likely why I prefer them, but inlining hardly matters with large or virtual methods since those usually aren't inlined anyways.

Reply:
`null is not a valid value for` might be better phrasing; looks good otherwise.

As an aside, performance of thrown exceptions is low on my list of things to consider in a library that should only (at least as far as possible) throw on programmer errors.

Reply:
:+1: